File: tests/python3/testSerialization.py
import strus
import os
import sys
import utils
from collections import namedtuple, OrderedDict

if len(sys.argv) > 1:
    outputdir = sys.argv[1]
else:
    outputdir = "."

ctx = strus.Context()

class Example:
    a = 123
    b = "holoholo"
    xyx = 123.456

    def __init__( self, a, b, xyz):
        self.a = a
        self.b = b
        self.xyz = xyz

class ExampleStruct:
    e1 = None
    e2 = None

    def __init__( self, e1, e2):
        self.e1 = e1
        self.e2 = e2

ExampleTuple = namedtuple("ExampleStruct", "field1 field2 field3")


example = Example( 1, "goloman", 12.45)
e1 = Example( 567, "hoho", 123.456)
e2 = Example( 789, "huhu", 456.789)
estruct1 = ExampleStruct( e1, e2)
estruct2 = ExampleStruct( example, estruct1)
estruct3 = OrderedDict((('foo', "bar"), ('spamm', "eggs"), ('ham', None)))
estruct4 = ExampleTuple(  99, 1.23, "blabla" )

result = "result:\n"
result += "(%u)\n%s\n" % (1,ctx.debug_serialize( example, True))
result += "(%u)\n%s\n" % (2,ctx.debug_serialize( estruct1, True))
result += "(%u)\n%s\n" % (3,ctx.debug_serialize( estruct2, True))
result += "(%u)\n%s\n" % (4,ctx.debug_serialize( estruct3, True))
result += "(%u)\n%s\n" % (5,ctx.debug_serialize( estruct4, True))

expected = """result:
(1)
a:1, b:"goloman", xyz:12.45
(2)
e1:{a:567, b:"hoho", xyz:123.456}, e2:{a:789, b:"huhu", xyz:456.789}
(3)
e1:{a:1, b:"goloman", xyz:12.45}, e2:{e1:{a:567, b:"hoho", xyz:123.456}, e2:{a:789, b:"huhu", xyz:456.789}}
(4)
foo:"bar", ham:NULL, spamm:"eggs"
(5)
99, 1.23, "blabla"
"""

utils.verifyTestOutput( outputdir, result, expected)

File: tests/lua/testSerialization.lua
--
-- This test is not part of the automated tests, because it does not run with a deterministic result and has to be verified manually
--
require "string"
require "math"
require "utils"
require "math"

local outputdir = arg[1] or '.'
local ctx = strus_Context.new( ctxconfig)

Example = {}
function Example.new( a, b, xyz )
	return {a=a, b=b, xyz=xyz}
end

ExampleStruct = {}
function ExampleStruct.new( e1, e2 )
	return {e1 = e1, e2 = e2}
end

example = Example.new( 1, "goloman", 12.45)
e1 = Example.new( 567, "hoho", 123.456)
e2 = Example.new( 789, "huhu", 456.789)
estruct1 = ExampleStruct.new( e1, e2)
estruct2 = ExampleStruct.new( example, estruct1)

result = "result:\n" ..
	string.format( "(%u)\n%s\n", 1, ctx:debug_serialize( example)) ..
	string.format( "(%u)\n%s\n", 2, ctx:debug_serialize( estruct1)) ..
	string.format( "(%u)\n%s\n", 3, ctx:debug_serialize( estruct2))

local expected = [[
result:
(1)
name b
value 'goloman'
name a
value 1
name xyz
value 12.45

(2)
name e2
open
  name b
  value 'huhu'
  name a
  value 789
  name xyz
  value 456.789
close
name e1
open
  name b
  value 'hoho'
  name a
  value 567
  name xyz
  value 123.456
close

(3)
name e2
open
  name e2
  open
    name b
    value 'huhu'
    name a
    value 789
    name xyz
    value 456.789
  close
  name e1
  open
    name b
    value 'hoho'
    name a
    value 567
    name xyz
    value 123.456
  close
close
name e1
open
  name b
  value 'goloman'
  name a
  value 1
  name xyz
  value 12.45
close
]]

verifyTestOutput( outputdir, result, expected)

